finality-aleph/src/data_io.rs

```diff
@@ -252,7 +272,28 @@ pub(crate) async fn refresh_best_chain<B: BlockT, SC: SelectChain<B>>(
                     .best_chain()
                     .await
                     .expect("No best chain");
-                *best_chain.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
+                if *new_best_header.number() <= max_block_num {
+                    *data.lock() = AlephData::new(new_best_header.hash(), *new_best_header.number());
+                } else {
+                    // we check if prev_best_header is an ancestor of new_best_header:
+                    if data.lock().number < max_block_num {
+                        let reduced_header = reduce_to_num(new_best_header.clone(), max_block_num);
+                        *data.lock() = AlephData::new(reduced_header.hash(), *reduced_header.number());
+                    } else {
+                        let is_ancestor = if let Some(prev_header) = prev_best_header {
```

Comment:
This whole thing seems unnecessary – replacing the best block by the latest within-session ancestor of the new best chain is what we want to do anyway, right? Unless I'm reading the code wrong this is what we end up doing anyway, only sometimes doing some unnecessary additional ancestor tracking.

Reply:
Actually, the whole if structure seems irrelevant – `reduce_to_num` is an approximate noop when `header.num() <= num`, so you can just always run it and replace with the result. This would make even the function unnecessary, since the code would be in only one place.

I am quite worried that I might be missing some big thing that makes all I wrote extremely stupid. xd

Reply:
The point why this is so complicated is that it optimizes for the case when `best_block` is much higher than `max_block_num`. The simplest thing would be to always go down via ancestors from `new_best_header` to `max_block_num`: however, this might be quite expensive (it's not clear to me whether this even sits in RAM, or maybe we need to read stuff from disk, if the gap is say 20k blocks)... So the optimization is to essentially remember the last call to `best_block` as `prev_best_header` and check is `new_best_header` extends the branch of `prev_best_header` -- if so, then we don't need to do the expensive backtracking again, because we did it in the last try...

If there is a simple, yet correct way to do than I would be happy to hear :P

Reply:
Ah, yes, now I see. >.< Thanks for the explanation.

Since this is a bit confusing maybe a comment with a slightly longer explanation would be useful. Although I am not very sure about that, I kinda feel stupid for not understanding it before, so it might not actually be necessary.

Reply:
Yeah, it's probably not easy to see this at first -- added a comment.